### PR TITLE
add secrets subcommand to osc

### PR DESCRIPTION
**Pull request by deads2k**
Adds 
 1. `osc secrets new` - same arguments as the existing bundle command, but it creates the secret instead of simply displaying it.
 1. `osc secrets new-dockercfg` - takes flags to create a valid dockercfg secret.

@fabianofranz ptal
@smarterclayton command names and are there others we should have (I'll make TODOs)

### CODE REVIEW

File: pkg/cmd/cli/secrets/new.go [hunk 1]
```diff
@@ -28,66 +31,53 @@ type CreateSecretOptions struct {
...
-	options := NewDefaultOptions()
+func NewCmdCreateSecret(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := NewCreateSecretOptions()
+	options.Out = out
 
 	cmd := &cobra.Command{
 		Use:   fmt.Sprintf("%s NAME SOURCE [SOURCE ...]", name),
-		Short: "Bundle files (or files within directories) into a Kubernetes secret",
-		Long: fmt.Sprintf(`Bundle files (or files within directories) into a Kubernetes secret.
+		Short: "Create a new Kubernetes secret based on a file or files within a directory",
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/new.go]: I think we can remove the Kubernetes term from the descriptions.

**Comment by derekwaynecarr** [comment]
I still think the ability to easily write a secret is useful.  For example, I think it will be more common to check a secret into source control than to create it on the fly direct to server.  Can we preserve an option that lets me bundle a secret locally instead of always writing to server.  I think of it similar to osc process and then piping into osc create.

**Comment by smarterclayton** [comment]
-o yaml

----- Original Message -----
> I still think the ability to easily write a secret is useful.  For example, I
> think it will be more common to check a secret into source control than to
> create it on the fly direct to server.  Can we preserve an option that lets
> me bundle a secret locally instead of always writing to server.  I think of
> it similar to osc process and then piping into osc create.
> 
> Sent from my iPhone
>

**Comment by deads2k** [comment]
> -o yaml

I removed that.  I'll add it back in.

File: hack/test-cmd.sh [hunk 2]
```diff
@@ -312,6 +312,16 @@ echo "templates: ok"
 [ "$(openshift admin TYPO; echo $? | grep '1')" ]
 [ "$(openshift cli TYPO; echo $? | grep '1')" ]
 [ "$(osc policy TYPO; echo $? | grep '1')" ]
+[ "$(osc secrets TYPO; echo $? | grep '1')" ]
+
+
+osc secrets new-dockercfg dockercfg --docker-username=sample-user --docker-password=sample-password --docker-email=fake@example.org
+osc describe secrets/dockercfg | grep "dockercfg:" | awk '{print $2}' > ${HOME}/.dockercfg
```
Comment (smarterclayton) [inline hack/test-cmd.sh]: Why to .dockercfg?

Comment (deads2k) [inline hack/test-cmd.sh]: > Why to .dockercfg?

The name has to match the filename the secret data map requires.

Comment (smarterclayton) [inline hack/test-cmd.sh]: Ah

File: pkg/cmd/cli/secrets/dockercfg.go [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+	Out io.Writer
+}
+
+// NewCmdCreateDockerConfigSecret creates a command object for making a dockercfg secret
+func NewCmdCreateDockerConfigSecret(name, fullName string, f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	o := &CreateDockerConfigOptions{Out: out}
+
+	cmd := &cobra.Command{
+		Use:   name + " <secret name>",
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/dockercfg.go]: There's a pattern for this now - check the other commands

Comment (smarterclayton) [inline pkg/cmd/cli/secrets/dockercfg.go]: Also need long descriptions

File: pkg/cmd/cli/secrets/subcommand.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+const SecretsRecommendedName = "secrets"
+
+func NewCmdSecrets(name, fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	// Parent command to which all subcommands are added.
+	cmds := &cobra.Command{
+		Use:   name,
+		Short: "Manage secrets",
+		Long:  `Manage secrets`,
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/subcommand.go]: Better long descriptions, explain secrets and what these commands can help users with

File: pkg/cmd/cli/secrets/new.go [hunk 5]
```diff
@@ -28,66 +31,53 @@ type CreateSecretOptions struct {
...
-			}
+			_, err := options.CreateSecret()
+			cmdutil.CheckErr(err)
 		},
 	}
 
 	cmd.Flags().BoolVarP(&options.Quiet, "quiet", "q", options.Quiet, "Suppress warnings")
 	cmd.Flags().VarP(&options.Sources, "source", "f", "List of filenames or directories to use as sources of Kubernetes Secret.Data")
 	cmd.Flags().StringVar(&options.SecretTypeName, "type", "", "The type of secret")
-	cmdutil.AddPrinterFlags(cmd)
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/new.go]: Need to keep this

File: pkg/cmd/cli/secrets/dockercfg.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+				cmdutil.CheckErr(err)
+			}
+
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Username, "docker-username", "", "username for docker registry authentication")
+	cmd.Flags().StringVar(&o.Password, "docker-password", "", "password for docker registry authentication")
+	cmd.Flags().StringVar(&o.EmailAddress, "docker-email", "", "email for docker registry")
+	cmd.Flags().StringVar(&o.RegistryLocation, "docker-server", "https://index.docker.io/v1/", "server location for docker registry")
```
Comment (smarterclayton) [inline pkg/cmd/cli/secrets/dockercfg.go]: Need to support -o

**Comment by mfojtik** [comment]
@deads2k @smarterclayton we have 'SourceSecret' in BuildConfig that has to take "$HOME/.ssh/id_rsa" file content to allow pulling from private GIT repositories... However that filename is not valid "key" format.... can we add optional argument to translate the filename to something else? (like `ssh-privatekey` in this case).

**Comment by smarterclayton** [comment]
Yeah - --key seems fine

**Comment by mfojtik** [comment]
@smarterclayton thanks! that will make the docs more consistent about creating secrets from build

**Comment by liggitt** [comment]
how does `--key` work with multiple files?

**Comment by mfojtik** [comment]
@liggitt how about `osc secret new private-key $HOME/.ssh/id_rsa:foo` (where the id_rsa will be translated to foo)

**Comment by smarterclayton** [comment]
Either has to be in order (if user lists files), or applies to all? if bulk.  Otherwise we have to invent a mapping format which is ugly.  What other alternatives?

**Comment by mfojtik** [comment]
@smarterclayton we will have to do mapping anyway as the secrets keys has to be DNS names (IMHO). So secret new will fail for all files that does not match that format.

**Comment by liggitt** [comment]
dns name requirement is headed toward being relaxed

**Comment by liggitt** [comment]
I don't like --key, but if we do it, we have to make sure the number of keys you give us matches the number of items we added

**Comment by liggitt** [comment]
also probably shouldn't allow directory listing as the input, since the order isn't deterministic

**Comment by mfojtik** [comment]
@liggitt can't we just do `--key-mappings=id_rsa:private-ssh-key,foo_bar:foobar`? Then the key will be translated prior to creation

**Comment by smarterclayton** [comment]
Hrm - really that's a different mode.  I think if we do that we don't take arguments.

Alternative - accept foo=bar syntax like labels and env.

**Comment by deads2k** [comment]
I'd like to make the `key-mappings` (or whatever its called) changes in a separate pull.

**Comment by smarterclayton** [comment]
That's find

**Comment by deads2k** [comment]
comments addressed.

**Comment by smarterclayton** [comment]
Change is approved once reviews are complete.

**Comment by deads2k** [comment]
more help added.

**Comment by deads2k** [comment]
@liggitt @fabianofranz ptal.

**Comment by deads2k** [comment]
@derekwaynecarr

**Comment by derekwaynecarr** [comment]
Sorry for the duplicate comments, I went back in commit history.

File: hack/test-cmd.sh [hunk 7]
```diff
@@ -320,6 +320,19 @@ echo "templates: ok"
 [ "$(openshift admin TYPO; echo $? | grep '1')" ]
 [ "$(openshift cli TYPO; echo $? | grep '1')" ]
 [ "$(oc policy TYPO; echo $? | grep '1')" ]
+[ "$(oc secrets TYPO; echo $? | grep '1')" ]
+
+
+oc secrets new-dockercfg dockercfg --docker-username=sample-user --docker-password=sample-password --docker-email=fake@example.org
+oc describe secrets/dockercfg | grep "dockercfg:" | awk '{print $2}' > ${HOME}/.dockercfg
```
Comment (derekwaynecarr) [inline hack/test-cmd.sh]: If you don't want to do a Go template now, put a ``` #TODO this is fragile, move to Go template ```

File: pkg/cmd/cli/secrets/dockercfg.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+)
+
+const (
+	CreateDockerConfigSecretRecommendedName = "new-dockercfg"
+
+	createDockercfgLong = `Create a new dockercfg secret
+
+Dockercfg secrets are used to authenticate against docker registries.
+
+When using plain docker, you can authenticate to a given registry by running
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/dockercfg.go]: @smarterclayton - do we need to capitalize Docker versus docker?  Is it only Docker when referring to the company and not the tool?

Comment (smarterclayton) [inline pkg/cmd/cli/secrets/dockercfg.go]: It should be capitalized when we refer to Docker the command line client.  Also, "plain docker" is awkward.  "When using the Docker command line to push images, you can ..."

----- Original Message -----

File: pkg/cmd/cli/secrets/subcommand.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+)
+
+const SecretsRecommendedName = "secrets"
+
+const (
+	secretsLong = `Manage secrets in your project.
+
+Secrets are used to store confidential information that should not be contained inside of an image.
+They are commonly used to hold things like keys for authentication to other internal systems like 
+docker registries.`
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/subcommand.go]: Docker registries

File: pkg/cmd/cli/secrets/dockercfg.go [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+			}
+
+			if err := o.CreateDockerSecret(); err != nil {
+				cmdutil.CheckErr(err)
+			}
+
+		},
+	}
+
+	cmd.Flags().StringVar(&o.Username, "docker-username", "", "username for docker registry authentication")
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/dockercfg.go]: Docker registry

File: pkg/cmd/cli/secrets/dockercfg.go [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+	cmd.Flags().StringVar(&o.Password, "docker-password", "", "password for docker registry authentication")
```
Comment (derekwaynecarr) [inline pkg/cmd/cli/secrets/dockercfg.go]: Docker registry, you get the idea.

**Comment by derekwaynecarr** [comment]
add the requested comment, fix docker -> Docker, then LGTM

**Comment by deads2k** [comment]
> add the requested comment, fix docker -> Docker, then LGTM

done

**Comment by deads2k** [comment]
[merge]

**Comment by smarterclayton** [comment]
Needs an upstream pull for the describer bit

**Comment by deads2k** [comment]
> Needs an upstream pull for the describer bit

Didn't go as smoothly as I hoped. https://github.com/GoogleCloudPlatform/kubernetes/pull/9321

**Pull request merged**